text/BartekJachowicz/opis.tex

```diff
@@ -0,0 +1,129 @@
+    \item $W(A(1:u), B(1:v)) = (m - p)/2$, gdzie $W(A, B) = |A| - |C|$
+\end{enumerate}
+\end{definition}
+
+Perfekcyjne cięcie pozwala podzielić zadany problem zgodnie z definicją na dwa podproblemy: \textit{LCS(A(1:u), B(1:v))} oraz \textit{LCS(A(u+1:m), B(v+1:n))}, a z otrzymanych wyników utworzyć rozwiązanie.
+
+Głównym problemem algorytmu jest zatem znalezienie perfekcyjnego cięcia. Do rozwiązania tego problemu wykorzystywane są dwie procedury: \textit{calmid} oraz \textit{fillone}. By lepiej zrozumieć ich ideę wprowadźmy najpierw następujące definicje:
+
+\begin{definition}{}
+$L_i(k)$ oznacza największe $h$ dla którego słowa \textit{A(i:m), B(h:n)} mają najdłuższy wspólny podciąg długości $k$.
+\end{definition}
+
+\begin{definition}
+$L_i^*(k)$ oznacza najmniejsze $h$ dla którego słowa \textit{A(i:m), B(h:n)} mają najdłuższy wspólny podciąg długości $k$.
+\end{definition}
+
+\noindent W pracy przedstawiona została zależność pomiędzy \textit{perfekcyjnym cięciem (u, v)} a wartościami $L_i(k)$ oraz $L_i^*(k)$. Pozwala ona znaleźć perfekcyjne cięcie słów $A$, $B$ jeśli zna się wartości jeśli zna się wartości $L_i(k)$ oraz $L_i^*(k)$ dla odpowiednich $i$, $k$:
+
+\begin{theorem}{}{perf-cut}
+Para (u, v) jest \textit{perfekcyjnym cięciem} wtedy i tylko wtedy gdy $L_{u+1}(m - u - w') \geq v \geq L^*_u(u - w)$, gdzie $w = \lfloor\frac{m-p}{2}\rfloor$ oraz $w' = \lceil\frac{m-p}{2}\rceil$.
+\end{theorem}
+
+Procedura \textit{calmid(A, B, m, n, x, LL, r)} oblicza wartości $L_{m-x+1-j}(j)$ za pomocą procedury \textit{fillone(A, B, m, n, R1, R2, r, s)} i zapisuje je w tablicy $LL$. Zmienna $r$ po wykonaniu funkcji oznacza natomiast największe j dla którego dana wartość jest poprawnie zdefiniowana. Procedura \textit{fillone(A, B, m, n, R1, R2, r, s)} używa dwóch tablic by obliczyć wartości $L_{s+1}(0), L_s(1), L_{s-1}(2), ...$. Obliczenia te wykonywane są iteracyjne, z wykorzystaniem poprzednich wartości. Poprawność tych funkcji oraz dokładne ich działanie opiszę przy dowodzie ich poprawności.
+
+
+\section{Opis dowodu poprawności oraz złożoności algorytmu}
+Prezentowany w pracy dowód poprawności algorytmu podzielony został na dowód poprawności dla każdej z procedur zaprezentowanych przez autorów. Na początek przedstawmy dowody poprawności dla procedur \textit{fillone} oraz \textit{calmid} których działanie jest kluczowe dla większości pozostałych funkcji algorytmu. Następnie przyjrzymy się metodzie znajdowania \textit{perfekcyjnego cięcia}. Pozostałe procedury działają w sposób dość jasny do zrozumienia.
+
+\subsection{Procedura \textit{fillone}}
+
+Tak jak zostało już wspomniane procedura \textit{fillone(A, B, m, n, $R_1, R_2$, r, s)} używa dwóch tablic by obliczyć wartości $L_{s+1}(0), L_s(1), L_{s-1}(2), ...$. Sposób w jaki uzyskiwane są kolejne wartości wynika z następującego lematu:
```

Comment:
OK, please add how you get the first value. I know, it's straightforward, but let's not force the reader to check the code for it.

Reply:
Done